solr/core/src/java/org/apache/solr/search/TopLevelJoinQuery.java

```diff
@@ -0,0 +1,218 @@
+      return new ConstantScoreWeight(this, boost) {
+        public Scorer scorer(LeafReaderContext context) throws IOException {
+          if (toBitsetBounds.lower == BitsetBounds.NO_MATCHES) {
+            return null;
+          }
+
+          final DocIdSetIterator toApproximation = (toMultivalued) ? context.reader().getSortedSetDocValues(toField) :
+              context.reader().getSortedDocValues(toField);
+          if (toApproximation == null) {
+            return null;
+          }
+
+          final int docBase = context.docBase;
+          return new ConstantScoreScorer(this, this.score(), scoreMode, new TwoPhaseIterator(toApproximation) {
+            public boolean matches() throws IOException {
+              final boolean hasDoc = topLevelToDocValues.advanceExact(docBase + approximation.docID());
+              if (hasDoc) {
+                for (long ord = topLevelToDocValues.nextOrd(); ord != -1L; ord = topLevelToDocValues.nextOrd()) {
+                  if (toOrdBitSet.get(ord)) {
+                    return true;
+                  }
+                }
+              }
+              return false;
+            }
+
+            public float matchCost() {
+              return 10.0F;
+            }
+          });
+
+        }
+
+        public boolean isCacheable(LeafReaderContext ctx) {
+          return false;
+        }
+      };
+    } catch (IOException e) {
+      throw new RuntimeException(e);
+    }
+  }
+
+  private Weight createNoMatchesWeight(float boost) {
+    return new ConstantScoreWeight(this, boost) {
+      @Override
+      public Scorer scorer(LeafReaderContext context) throws IOException {
+        return null;
+      }
+
+      @Override
+      public boolean isCacheable(LeafReaderContext ctx) {
+        return false;
+      }
+    };
+  }
+
+  private SortedSetDocValues validateAndFetchDocValues(SolrIndexSearcher solrSearcher, String fieldName, String querySide) throws IOException {
+    final IndexSchema schema = solrSearcher.getSchema();
+    final SchemaField field = schema.getFieldOrNull(fieldName);
+    if (field == null) {
+      throw new SolrException(SolrException.ErrorCode.BAD_REQUEST, querySide + " field '" + fieldName + "' does not exist");
+    }
+
+    if (!field.hasDocValues()) {
+      throw new SolrException(SolrException.ErrorCode.BAD_REQUEST,
+          "'top-level' join queries require both 'from' and 'to' fields to have docValues, but " + querySide +
+              " field [" + fieldName +  "] does not.");
+    }
+
+    final LeafReader leafReader = solrSearcher.getSlowAtomicReader();
+    if (field.multiValued()) {
+      return DocValues.getSortedSet(leafReader, fieldName);
+    }
+    return DocValues.singleton(DocValues.getSorted(leafReader, fieldName));
+  }
+
+  private LongBitSet findOrdinalsMatchingFromQuery(SolrIndexSearcher fromSearcher, SortedSetDocValues fromDocValues) throws IOException {
```

Comment:
Couldn't you rename the method and it's variables to remove "From" and it would be just as valid?  Okay I see why not; this is an instance method that refers to 'fromField' but if that's the only distinction then maybe it should be a static method and take that argument?